lusidtools/extract/group_holdings.py

```diff
@@ -0,0 +1,364 @@
+    if group_by_portfolio:
+        if len(set(holdings_to_join.keys())) != len(holdings_to_join.keys()):
+            raise ValueError("There are duplicate Portfolios in the Portfolio Group. This is not currently supported. "
+                             "Please ensure that there are no duplicates in your Porfolio Group via sub-groups and try again")
+        return holdings_to_join
+
+    # Get the list of holdings themselves from each VersionedResourceListOfPortfolioHolding
+    holding_lists = [holdings.values for holdings in holdings_to_join.values()]
+    # Flatten the holdings into a single list
+    all_holdings = [holding for holding_list in holding_lists for holding in holding_list]
+    # Initialise a dictionary to hold the positions against a key
+    all_holdings_keyed = {}
+    # Construct the keyed dictionary using the LUSID instrument id, holding type and currency as the unique key for each holding
+    [all_holdings_keyed.setdefault(f"{holding.instrument_uid}-{holding.holding_type}-{holding.cost.currency}", []).append(
+        holding
+    ) for holding in all_holdings]
+
+    # Initialise a list to hold the joined holdings
+    holdings = []
+
+    # Reduce the list of holdings against each key to a single holding
+    for key, value in all_holdings_keyed.items():
+        holdings.append(lusid.models.PortfolioHolding(
+            # Use the first instrument_uid (could be picked off the key)
+            instrument_uid=value[0].instrument_uid,
+            # Use the first holding type (could be picked off the key)
+            holding_type=value[0].holding_type,
+            # Add the metric fields together
+            units=reduce((lambda x, y: x+y), list(map(lambda x: x.units, value))),
+            settled_units=reduce((lambda x, y: x+y), list(map(lambda x: x.settled_units, value))),
+            cost=lusid.models.CurrencyAndAmount(
+                # Use the first currency (could be picked off the key)
+                currency=value[0].cost.currency,
+                amount=reduce((lambda x, y: x+y), list(map(lambda x: x.cost.amount, value)))
+            ),
+            cost_portfolio_ccy=lusid.models.CurrencyAndAmount(
+                # Use the first currency, these holdings could be from different portfolios, so the validity of this is questionable
+                currency=value[0].cost_portfolio_ccy.currency,
+                amount=reduce((lambda x,y: x+y), list(map(lambda x: x.cost_portfolio_ccy.amount, value)))
+            ),
+            # Takes the properties from the first value, only allows instrument properties
+            properties={
+                property_key: value for property_key, value in value[0].properties.items() if value.key.split("/")[0] == "Instrument"
+            }
+        )
+    )
+
+    # Use the other details of the first VersionedResourceListOfPortfolioHolding in the response e.g. links etc.
+    joined_holdings = list(holdings_to_join.values())[0]
```

Comment:
this will throw an exception if an empty dict is passed in

Reply:
Removed due to using a list of PortfolioHolding instead